File: c_programs_and_containers/build_blocks/knowledge_base/mqtt/job_queue/src/mqtt_queue.c
/**
 * mqtt_queue.c - MQTT Queue Publisher & Reader implementation
 *
 * Wraps libmosquitto with thread-safe connect/subscribe/publish operations
 * and a persistent-session reader for offline message queuing.
 */

#define _POSIX_C_SOURCE 200809L   /* clock_gettime, nanosleep, strdup */

#include "mqtt_queue.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <time.h>

/* Suppress unused-parameter warnings for mosquitto callbacks */
#define UNUSED(x) (void)(x)

/* ================================================================== */
/*  Helpers                                                            */
/* ================================================================== */

static char *safe_strdup(const char *s)
{
    return s ? strdup(s) : NULL;
}

/**
 * Wait on a condition variable with an absolute-time timeout.
 * Returns 0 if signalled, ETIMEDOUT on timeout, or other errno.
 */
static int cond_timedwait_ms(pthread_cond_t *cond, pthread_mutex_t *mtx, int ms)
{
    struct timespec ts;
    clock_gettime(CLOCK_REALTIME, &ts);
    ts.tv_sec  += ms / 1000;
    ts.tv_nsec += (ms % 1000) * 1000000L;
    if (ts.tv_nsec >= 1000000000L) {
        ts.tv_sec  += 1;
        ts.tv_nsec -= 1000000000L;
    }
    return pthread_cond_timedwait(cond, mtx, &ts);
}

/* ================================================================== */
/*  Message list                                                       */
/* ================================================================== */

void mqtt_msg_list_free(mqtt_msg_t *head)
{
    while (head) {
        mqtt_msg_t *next = head->next;
        free(head->topic);
        free(head->payload);
        free(head);
        head = next;
    }
}

static mqtt_msg_t *mqtt_msg_new(const char *topic, const void *payload, int payloadlen)
{
    mqtt_msg_t *m = calloc(1, sizeof(*m));
    if (!m) return NULL;

    m->topic   = safe_strdup(topic);
    m->payload = malloc((size_t)payloadlen + 1);
    if (m->payload) {
        memcpy(m->payload, payload, (size_t)payloadlen);
        m->payload[payloadlen] = '\0';
    }
    m->next = NULL;
    return m;
}

/* ================================================================== */
/*  Library init / cleanup                                             */
/* ================================================================== */

void mqtt_queue_lib_init(void)
{
    mosquitto_lib_init();
}

void mqtt_queue_lib_cleanup(void)
{
    mosquitto_lib_cleanup();
}

/* ================================================================== */
/*  QueuePublisher                                                     */
/* ================================================================== */

/* ---- callbacks --------------------------------------------------- */

static void pub_on_connect(struct mosquitto *mosq, void *userdata, int rc)
{
    UNUSED(mosq);
    mqtt_publisher_t *pub = (mqtt_publisher_t *)userdata;
    pthread_mutex_lock(&pub->lock);
    pub->connected = (rc == 0);
    printf("[publisher] on_connect rc=%d (%s)\n", rc,
           rc == 0 ? "success" : mosquitto_connack_string(rc));
    pthread_cond_signal(&pub->connect_cond);
    pthread_mutex_unlock(&pub->lock);
}

static void pub_on_disconnect(struct mosquitto *mosq, void *userdata, int rc)
{
    UNUSED(mosq);
    mqtt_publisher_t *pub = (mqtt_publisher_t *)userdata;
    pthread_mutex_lock(&pub->lock);
    pub->connected = false;
    printf("[publisher] on_disconnect rc=%d\n", rc);
    pthread_cond_signal(&pub->connect_cond);
    pthread_mutex_unlock(&pub->lock);
}

/* ---- API --------------------------------------------------------- */

int mqtt_publisher_init(mqtt_publisher_t *pub, const mqtt_queue_config_t *cfg)
{
    memset(pub, 0, sizeof(*pub));
    pub->cfg = *cfg;

    pthread_mutex_init(&pub->lock, NULL);
    pthread_cond_init(&pub->connect_cond, NULL);

    pub->mosq = mosquitto_new(cfg->client_id, cfg->clean_session, pub);
    if (!pub->mosq) {
        fprintf(stderr, "[publisher] mosquitto_new failed: %s\n", strerror(errno));
        return -1;
    }

    if (cfg->username) {
        mosquitto_username_pw_set(pub->mosq, cfg->username, cfg->password);
    }

    mosquitto_connect_callback_set(pub->mosq, pub_on_connect);
    mosquitto_disconnect_callback_set(pub->mosq, pub_on_disconnect);

    return 0;
}

int mqtt_publisher_connect(mqtt_publisher_t *pub, int timeout_ms)
{
    int rc = mosquitto_connect(pub->mosq, pub->cfg.host, pub->cfg.port, pub->cfg.keepalive);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[publisher] mosquitto_connect: %s\n", mosquitto_strerror(rc));
        return -1;
    }

    rc = mosquitto_loop_start(pub->mosq);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[publisher] mosquitto_loop_start: %s\n", mosquitto_strerror(rc));
        return -1;
    }

    /* Wait for CONNACK */
    pthread_mutex_lock(&pub->lock);
    if (!pub->connected) {
        cond_timedwait_ms(&pub->connect_cond, &pub->lock, timeout_ms);
    }
    bool ok = pub->connected;
    pthread_mutex_unlock(&pub->lock);

    if (!ok) {
        fprintf(stderr, "[publisher] connect timeout\n");
        mosquitto_loop_stop(pub->mosq, true);
        return -1;
    }
    return 0;
}

void mqtt_publisher_disconnect(mqtt_publisher_t *pub)
{
    mosquitto_disconnect(pub->mosq);
    mosquitto_loop_stop(pub->mosq, false);
    pthread_mutex_lock(&pub->lock);
    pub->connected = false;
    pthread_mutex_unlock(&pub->lock);
}

void mqtt_publisher_destroy(mqtt_publisher_t *pub)
{
    if (pub->mosq) {
        mosquitto_destroy(pub->mosq);
        pub->mosq = NULL;
    }
    pthread_mutex_destroy(&pub->lock);
    pthread_cond_destroy(&pub->connect_cond);
}

int mqtt_publisher_publish(mqtt_publisher_t *pub,
                           const char *topic,
                           const char *json_payload,
                           int qos,
                           bool retain)
{
    int mid = 0;
    int rc = mosquitto_publish(pub->mosq, &mid, topic,
                               (int)strlen(json_payload), json_payload,
                               qos, retain);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[publisher] publish failed: %s\n", mosquitto_strerror(rc));
        return -1;
    }
    return 0;
}

int mqtt_publisher_publish_batch(mqtt_publisher_t *pub,
                                 const char *topic,
                                 const char **json_payloads,
                                 int count,
                                 int qos,
                                 bool retain,
                                 int delay_between_ms)
{
    int ok = 0;
    for (int i = 0; i < count; i++) {
        if (mqtt_publisher_publish(pub, topic, json_payloads[i], qos, retain) == 0) {
            ok++;
            printf("  [batch] published %d/%d\n", i + 1, count);
        } else {
            fprintf(stderr, "  [batch] failed %d/%d\n", i + 1, count);
        }
        if (delay_between_ms > 0 && i < count - 1) {
            struct timespec ts = {
                .tv_sec  = delay_between_ms / 1000,
                .tv_nsec = (delay_between_ms % 1000) * 1000000L
            };
            nanosleep(&ts, NULL);
        }
    }
    return ok;
}

bool mqtt_publisher_is_connected(const mqtt_publisher_t *pub)
{
    return pub->connected;
}

/* ================================================================== */
/*  QueueReader                                                        */
/* ================================================================== */

/* ---- callbacks --------------------------------------------------- */

static void rdr_on_connect(struct mosquitto *mosq, void *userdata, int rc)
{
    UNUSED(mosq);
    mqtt_reader_t *rdr = (mqtt_reader_t *)userdata;
    pthread_mutex_lock(&rdr->lock);

    rdr->connected = (rc == 0);

    printf("[reader] on_connect rc=%d (%s)\n", rc,
           rc == 0 ? "success" : mosquitto_connack_string(rc));

    pthread_cond_signal(&rdr->connect_cond);
    pthread_mutex_unlock(&rdr->lock);
}

static void rdr_on_disconnect(struct mosquitto *mosq, void *userdata, int rc)
{
    UNUSED(mosq);
    mqtt_reader_t *rdr = (mqtt_reader_t *)userdata;
    pthread_mutex_lock(&rdr->lock);
    rdr->connected = false;
    printf("[reader] on_disconnect rc=%d\n", rc);
    pthread_cond_signal(&rdr->connect_cond);
    pthread_mutex_unlock(&rdr->lock);
}

static void rdr_on_message(struct mosquitto *mosq, void *userdata,
                           const struct mosquitto_message *msg)
{
    UNUSED(mosq);
    mqtt_reader_t *rdr = (mqtt_reader_t *)userdata;
    mqtt_msg_t *m = mqtt_msg_new(msg->topic, msg->payload, msg->payloadlen);
    if (!m) return;

    pthread_mutex_lock(&rdr->lock);
    if (rdr->msg_tail) {
        rdr->msg_tail->next = m;
    } else {
        rdr->msg_head = m;
    }
    rdr->msg_tail = m;
    rdr->msg_count++;
    pthread_mutex_unlock(&rdr->lock);
}

static void rdr_on_subscribe(struct mosquitto *mosq, void *userdata,
                             int mid, int qos_count, const int *granted_qos)
{
    UNUSED(mosq); UNUSED(granted_qos);
    mqtt_reader_t *rdr = (mqtt_reader_t *)userdata;
    printf("[reader] on_subscribe mid=%d qos_count=%d\n", mid, qos_count);

    pthread_mutex_lock(&rdr->lock);
    if (mid == rdr->subscribe_mid) {
        rdr->subscribe_ack = true;
        pthread_cond_signal(&rdr->subscribe_cond);
    }
    pthread_mutex_unlock(&rdr->lock);
}

/* ---- API --------------------------------------------------------- */

int mqtt_reader_init(mqtt_reader_t *rdr, const mqtt_queue_config_t *cfg)
{
    memset(rdr, 0, sizeof(*rdr));
    rdr->cfg = *cfg;

    pthread_mutex_init(&rdr->lock, NULL);
    pthread_cond_init(&rdr->connect_cond, NULL);
    pthread_cond_init(&rdr->subscribe_cond, NULL);

    rdr->mosq = mosquitto_new(cfg->client_id, cfg->clean_session, rdr);
    if (!rdr->mosq) {
        fprintf(stderr, "[reader] mosquitto_new failed: %s\n", strerror(errno));
        return -1;
    }

    if (cfg->username) {
        mosquitto_username_pw_set(rdr->mosq, cfg->username, cfg->password);
    }

    mosquitto_connect_callback_set(rdr->mosq, rdr_on_connect);
    mosquitto_disconnect_callback_set(rdr->mosq, rdr_on_disconnect);
    mosquitto_message_callback_set(rdr->mosq, rdr_on_message);
    mosquitto_subscribe_callback_set(rdr->mosq, rdr_on_subscribe);

    return 0;
}

int mqtt_reader_connect(mqtt_reader_t *rdr, int timeout_ms)
{
    int rc = mosquitto_connect(rdr->mosq, rdr->cfg.host, rdr->cfg.port, rdr->cfg.keepalive);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[reader] mosquitto_connect: %s\n", mosquitto_strerror(rc));
        return -1;
    }

    rc = mosquitto_loop_start(rdr->mosq);
    if (rc != MOSQ_ERR_SUCCESS) {
        fprintf(stderr, "[reader] mosquitto_loop_start: %s\n", mosquitto_strerror(rc));
        return -1;
    }

    pthread_mutex_lock(&rdr->lock);
    if (!rdr->connected) {
        cond_timedwait_ms(&rdr->connect_cond, &rdr->lock, timeout_ms);
    }
    bool ok = rdr->connected;
    pthread_mutex_unlock(&rdr->lock);

    if (!ok) {
        fprintf(stderr, "[reader] connect timeout\n");
        mosquitto_loop_stop(rdr->mosq, true);
        return -1;
    }
    return 0;
}

void mqtt_reader_disconnect(mqtt_reader_t *rdr)
{
    mosquitto_disconnect(rdr->mosq);
    mosquitto_loop_stop(rdr->mosq, false);
    pthread_mutex_lock(&rdr->lock);
    rdr->connected = false;
    pthread_mutex_unlock(&rdr->lock);
}

void mqtt_reader_destroy(mqtt_reader_t *rdr)
{
    /* Free any remaining messages */
    mqtt_msg_list_free(rdr->msg_head);
    rdr->msg_head = rdr->msg_tail = NULL;
    rdr->msg_count = 0;

    if (rdr->mosq) {
        mosquitto_destroy(rdr->mosq);
        rdr->mosq = NULL;
    }
    pthread_mutex_destroy(&rdr->lock);
    pthread_cond_destroy(&rdr->connect_cond);
    pthread_cond_destroy(&rdr->subscribe_cond);
}

int mqtt_reader_subscribe(mqtt_reader_t *rdr, const char *topic, int qos, int timeout_ms)
{
    pthread_mutex_lock(&rdr->lock);
    if (!rdr->connected) {
        pthread_mutex_unlock(&rdr->lock);
        fprintf(stderr, "[reader] subscribe: not connected\n");
        return -1;
    }

    rdr->subscribe_ack = false;
    int mid = 0;
    int rc = mosquitto_subscribe(rdr->mosq, &mid, topic, qos);
    if (rc != MOSQ_ERR_SUCCESS) {
        pthread_mutex_unlock(&rdr->lock);
        fprintf(stderr, "[reader] subscribe failed: %s\n", mosquitto_strerror(rc));
        return -1;
    }
    rdr->subscribe_mid = mid;

    /* Wait for SUBACK */
    while (!rdr->subscribe_ack) {
        int err = cond_timedwait_ms(&rdr->subscribe_cond, &rdr->lock, timeout_ms);
        if (err == ETIMEDOUT) {
            pthread_mutex_unlock(&rdr->lock);
            fprintf(stderr, "[reader] subscribe timeout waiting for SUBACK\n");
            return -1;
        }
    }
    pthread_mutex_unlock(&rdr->lock);
    return 0;
}

mqtt_msg_t *mqtt_reader_read_queue(mqtt_reader_t *rdr,
                                   const char *topic,
                                   int qos,
                                   int timeout_ms,
                                   int *out_count)
{
    /* Clear any old messages */
    mqtt_reader_clear(rdr);

    /* Subscribe if this isn't a resumed persistent session */
    if (!rdr->session_present) {
        if (mqtt_reader_subscribe(rdr, topic, qos, 2000) != 0) {
            if (out_count) *out_count = 0;
            return NULL;
        }
    } else {
        printf("[reader] using existing subscription from persistent session\n");
    }

    /* Collect messages for the specified timeout */
    struct timespec ts = {
        .tv_sec  = timeout_ms / 1000,
        .tv_nsec = (timeout_ms % 1000) * 1000000L
    };
    nanosleep(&ts, NULL);

    /* Detach the message list */
    pthread_mutex_lock(&rdr->lock);
    mqtt_msg_t *head = rdr->msg_head;
    int count = rdr->msg_count;
    rdr->msg_head = rdr->msg_tail = NULL;
    rdr->msg_count = 0;
    pthread_mutex_unlock(&rdr->lock);

    if (out_count) *out_count = count;
    return head;
}

void mqtt_reader_clear(mqtt_reader_t *rdr)
{
    pthread_mutex_lock(&rdr->lock);
    mqtt_msg_list_free(rdr->msg_head);
    rdr->msg_head = rdr->msg_tail = NULL;
    rdr->msg_count = 0;
    pthread_mutex_unlock(&rdr->lock);
}

bool mqtt_reader_is_connected(const mqtt_reader_t *rdr)
{
    return rdr->connected;
}

File: zig_programs_and_containers/building_blocks/knowledge_base/mqtt/job_queue/src/mqtt_queue.zig
//! mqtt_queue.zig - MQTT Queue Publisher & Reader Library
//!
//! Reliable queued messaging over MQTT v3.1.1 using persistent sessions.
//! Wraps the Mosquitto C client library (libmosquitto).
//!
//! Features:
//!   - QueuePublisher: publish JSON messages (single + batch) with QoS 1/2
//!   - QueueReader:    persistent-session consumer with offline message queuing
//!
//! Thread safety: each instance is internally synchronised via std.Thread.Mutex.

const std = @import("std");
const log = std.log.scoped(.mqtt_queue);

const c = @cImport({
    @cInclude("mosquitto.h");
});

// ──────────────────────────────────────────────────────────────────────
//  Public types
// ──────────────────────────────────────────────────────────────────────

pub const Config = struct {
    host: [:0]const u8 = "localhost",
    port: u16 = 1883,
    client_id: ?[:0]const u8 = null,
    keepalive: u16 = 60,
    username: ?[:0]const u8 = null,
    password: ?[:0]const u8 = null,
    clean_session: bool = true,
};

pub const Qos = enum(c_int) {
    at_most_once = 0,
    at_least_once = 1,
    exactly_once = 2,
};

/// A received MQTT message.
pub const Message = struct {
    topic: []u8,
    payload: []u8,
    next: ?*Message = null,

    /// Free a single message node (and its owned slices).
    pub fn deinit(self: *Message, allocator: std.mem.Allocator) void {
        allocator.free(self.topic);
        allocator.free(self.payload);
        allocator.destroy(self);
    }
};

/// Free an entire linked list of messages.
pub fn messageFreeList(head: ?*Message, allocator: std.mem.Allocator) void {
    var cur = head;
    while (cur) |node| {
        const next = node.next;
        node.deinit(allocator);
        cur = next;
    }
}

pub const Error = error{
    MosquittoNew,
    ConnectFailed,
    LoopStartFailed,
    ConnectTimeout,
    NotConnected,
    SubscribeFailed,
    SubscribeTimeout,
    PublishFailed,
    OutOfMemory,
};

// ──────────────────────────────────────────────────────────────────────
//  Library-level init / cleanup
// ──────────────────────────────────────────────────────────────────────

/// Call before any other mqtt_queue function (once per process).
pub fn libInit() void {
    _ = c.mosquitto_lib_init();
}

/// Call at program exit.
pub fn libCleanup() void {
    _ = c.mosquitto_lib_cleanup();
}

// ──────────────────────────────────────────────────────────────────────
//  QueuePublisher
// ──────────────────────────────────────────────────────────────────────

pub const Publisher = struct {
    mosq: *c.struct_mosquitto = undefined,
    cfg: Config,
    connected: bool = false,
    mutex: std.Thread.Mutex = .{},
    connect_cond: std.Thread.Condition = .{},

    const Self = @This();

    /// Initialise a publisher. Returns error on failure.
    pub fn init(cfg: Config) Error!Self {
        var self = Self{
            .cfg = cfg,
        };

        const cid: ?[*:0]const u8 = if (cfg.client_id) |id| id.ptr else null;
        const mosq = c.mosquitto_new(cid, cfg.clean_session, null);
        if (mosq == null) {
            log.err("[publisher] mosquitto_new failed", .{});
            return Error.MosquittoNew;
        }
        self.mosq = mosq.?;

        if (cfg.username) |user| {
            const pw: ?[*:0]const u8 = if (cfg.password) |p| p.ptr else null;
            _ = c.mosquitto_username_pw_set(self.mosq, user.ptr, pw);
        }

        c.mosquitto_connect_callback_set(self.mosq, pubOnConnect);
        c.mosquitto_disconnect_callback_set(self.mosq, pubOnDisconnect);

        return self;
    }

    /// Connect to broker. Blocks up to timeout_ms for CONNACK.
    pub fn connect(self: *Self, timeout_ms: u64) Error!void {
        // Set userdata to the caller's stable pointer (not the temporary from init)
        c.mosquitto_user_data_set(self.mosq, @ptrCast(self));

        const rc_conn = c.mosquitto_connect(
            self.mosq,
            self.cfg.host.ptr,
            @intCast(self.cfg.port),
            @intCast(self.cfg.keepalive),
        );
        if (rc_conn != c.MOSQ_ERR_SUCCESS) {
            log.err("[publisher] mosquitto_connect: {s}", .{c.mosquitto_strerror(rc_conn)});
            return Error.ConnectFailed;
        }

        const rc_loop = c.mosquitto_loop_start(self.mosq);
        if (rc_loop != c.MOSQ_ERR_SUCCESS) {
            log.err("[publisher] mosquitto_loop_start: {s}", .{c.mosquitto_strerror(rc_loop)});
            return Error.LoopStartFailed;
        }

        // Wait for CONNACK
        self.mutex.lock();
        defer self.mutex.unlock();

        if (!self.connected) {
            self.connect_cond.timedWait(&self.mutex, timeout_ms * std.time.ns_per_ms) catch {};
        }

        if (!self.connected) {
            log.err("[publisher] connect timeout", .{});
            _ = c.mosquitto_loop_stop(self.mosq, true);
            return Error.ConnectTimeout;
        }
    }

    /// Disconnect from broker.
    pub fn disconnect(self: *Self) void {
        _ = c.mosquitto_disconnect(self.mosq);
        _ = c.mosquitto_loop_stop(self.mosq, false);
        self.mutex.lock();
        defer self.mutex.unlock();
        self.connected = false;
    }

    /// Destroy publisher and free mosquitto resources.
    pub fn deinit(self: *Self) void {
        c.mosquitto_destroy(self.mosq);
    }

    /// Publish a single payload to topic. Returns error on failure.
    pub fn publish(
        self: *Self,
        topic: [:0]const u8,
        payload: []const u8,
        qos: Qos,
        retain: bool,
    ) Error!void {
        var mid: c_int = 0;
        const rc = c.mosquitto_publish(
            self.mosq,
            &mid,
            topic.ptr,
            @intCast(payload.len),
            payload.ptr,
            @intFromEnum(qos),
            retain,
        );
        if (rc != c.MOSQ_ERR_SUCCESS) {
            log.err("[publisher] publish failed: {s}", .{c.mosquitto_strerror(rc)});
            return Error.PublishFailed;
        }
    }

    /// Publish a batch of payloads. Returns the number successfully published.
    pub fn publishBatch(
        self: *Self,
        topic: [:0]const u8,
        payloads: []const []const u8,
        qos: Qos,
        retain: bool,
        delay_between_ms: u64,
    ) usize {
        var ok: usize = 0;
        for (payloads, 0..) |payload, i| {
            if (self.publish(topic, payload, qos, retain)) |_| {
                ok += 1;
                log.info("  [batch] published {d}/{d}", .{ i + 1, payloads.len });
            } else |_| {
                log.err("  [batch] failed {d}/{d}", .{ i + 1, payloads.len });
            }
            if (delay_between_ms > 0 and i < payloads.len - 1) {
                std.time.sleep(delay_between_ms * std.time.ns_per_ms);
            }
        }
        return ok;
    }

    pub fn isConnected(self: *const Self) bool {
        return self.connected;
    }

    // ── Mosquitto C callbacks (called from the network thread) ───────

    fn pubOnConnect(mosq: ?*c.struct_mosquitto, userdata: ?*anyopaque, rc: c_int) callconv(.C) void {
        _ = mosq;
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.mutex.lock();
        defer self.mutex.unlock();
        self.connected = (rc == 0);
        log.info("[publisher] on_connect rc={d}", .{rc});
        self.connect_cond.signal();
    }

    fn pubOnDisconnect(mosq: ?*c.struct_mosquitto, userdata: ?*anyopaque, rc: c_int) callconv(.C) void {
        _ = mosq;
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.mutex.lock();
        defer self.mutex.unlock();
        self.connected = false;
        log.info("[publisher] on_disconnect rc={d}", .{rc});
        self.connect_cond.signal();
    }
};

// ──────────────────────────────────────────────────────────────────────
//  QueueReader
// ──────────────────────────────────────────────────────────────────────

pub const Reader = struct {
    mosq: *c.struct_mosquitto = undefined,
    cfg: Config,
    allocator: std.mem.Allocator,
    connected: bool = false,
    session_present: bool = false,

    // Incoming message buffer (linked list, guarded by mutex)
    msg_head: ?*Message = null,
    msg_tail: ?*Message = null,
    msg_count: usize = 0,

    // Synchronisation
    mutex: std.Thread.Mutex = .{},
    connect_cond: std.Thread.Condition = .{},
    subscribe_cond: std.Thread.Condition = .{},
    subscribe_ack: bool = false,
    subscribe_mid: c_int = 0,

    const Self = @This();

    /// Initialise a reader. The allocator is used for incoming message buffers.
    pub fn init(cfg: Config, allocator: std.mem.Allocator) Error!Self {
        var self = Self{
            .cfg = cfg,
            .allocator = allocator,
        };

        const cid: ?[*:0]const u8 = if (cfg.client_id) |id| id.ptr else null;
        const mosq = c.mosquitto_new(cid, cfg.clean_session, null);
        if (mosq == null) {
            log.err("[reader] mosquitto_new failed", .{});
            return Error.MosquittoNew;
        }
        self.mosq = mosq.?;

        if (cfg.username) |user| {
            const pw: ?[*:0]const u8 = if (cfg.password) |p| p.ptr else null;
            _ = c.mosquitto_username_pw_set(self.mosq, user.ptr, pw);
        }

        c.mosquitto_connect_callback_set(self.mosq, rdrOnConnect);
        c.mosquitto_disconnect_callback_set(self.mosq, rdrOnDisconnect);
        c.mosquitto_message_callback_set(self.mosq, rdrOnMessage);
        c.mosquitto_subscribe_callback_set(self.mosq, rdrOnSubscribe);

        return self;
    }

    /// Connect to broker. Blocks up to timeout_ms for CONNACK.
    pub fn connect(self: *Self, timeout_ms: u64) Error!void {
        // Set userdata to the caller's stable pointer (not the temporary from init)
        c.mosquitto_user_data_set(self.mosq, @ptrCast(self));

        const rc_conn = c.mosquitto_connect(
            self.mosq,
            self.cfg.host.ptr,
            @intCast(self.cfg.port),
            @intCast(self.cfg.keepalive),
        );
        if (rc_conn != c.MOSQ_ERR_SUCCESS) {
            log.err("[reader] mosquitto_connect: {s}", .{c.mosquitto_strerror(rc_conn)});
            return Error.ConnectFailed;
        }

        const rc_loop = c.mosquitto_loop_start(self.mosq);
        if (rc_loop != c.MOSQ_ERR_SUCCESS) {
            log.err("[reader] mosquitto_loop_start: {s}", .{c.mosquitto_strerror(rc_loop)});
            return Error.LoopStartFailed;
        }

        self.mutex.lock();
        defer self.mutex.unlock();

        if (!self.connected) {
            self.connect_cond.timedWait(&self.mutex, timeout_ms * std.time.ns_per_ms) catch {};
        }

        if (!self.connected) {
            log.err("[reader] connect timeout", .{});
            _ = c.mosquitto_loop_stop(self.mosq, true);
            return Error.ConnectTimeout;
        }
    }

    /// Disconnect from broker.
    pub fn disconnect(self: *Self) void {
        _ = c.mosquitto_disconnect(self.mosq);
        _ = c.mosquitto_loop_stop(self.mosq, false);
        self.mutex.lock();
        defer self.mutex.unlock();
        self.connected = false;
    }

    /// Destroy reader and free all resources including buffered messages.
    pub fn deinit(self: *Self) void {
        self.clear();
        c.mosquitto_destroy(self.mosq);
    }

    /// Subscribe to a topic and block until SUBACK or timeout.
    pub fn subscribe(self: *Self, topic: [:0]const u8, qos: Qos, timeout_ms: u64) Error!void {
        self.mutex.lock();
        defer self.mutex.unlock();

        if (!self.connected) {
            log.err("[reader] subscribe: not connected", .{});
            return Error.NotConnected;
        }

        self.subscribe_ack = false;
        var mid: c_int = 0;
        const rc = c.mosquitto_subscribe(self.mosq, &mid, topic.ptr, @intFromEnum(qos));
        if (rc != c.MOSQ_ERR_SUCCESS) {
            log.err("[reader] subscribe failed: {s}", .{c.mosquitto_strerror(rc)});
            return Error.SubscribeFailed;
        }
        self.subscribe_mid = mid;

        // Wait for SUBACK
        while (!self.subscribe_ack) {
            self.subscribe_cond.timedWait(&self.mutex, timeout_ms * std.time.ns_per_ms) catch {
                log.err("[reader] subscribe timeout waiting for SUBACK", .{});
                return Error.SubscribeTimeout;
            };
        }
    }

    /// Collect messages for up to timeout_ms milliseconds.
    /// If the session was already present, skips re-subscribing.
    /// Returns a linked list of messages (caller must free with messageFreeList).
    pub fn readQueue(
        self: *Self,
        topic: [:0]const u8,
        qos: Qos,
        timeout_ms: u64,
        out_count: ?*usize,
    ) Error!?*Message {
        // Clear any old messages
        self.clear();

        // Subscribe if this isn't a resumed persistent session
        if (!self.session_present) {
            try self.subscribe(topic, qos, 2000);
        } else {
            log.info("[reader] using existing subscription from persistent session", .{});
        }

        // Collect messages for the specified timeout
        std.time.sleep(timeout_ms * std.time.ns_per_ms);

        // Detach the message list
        self.mutex.lock();
        defer self.mutex.unlock();

        const head = self.msg_head;
        const count = self.msg_count;
        self.msg_head = null;
        self.msg_tail = null;
        self.msg_count = 0;

        if (out_count) |cnt| {
            cnt.* = count;
        }
        return head;
    }

    /// Clear any buffered messages.
    pub fn clear(self: *Self) void {
        self.mutex.lock();
        const head = self.msg_head;
        self.msg_head = null;
        self.msg_tail = null;
        self.msg_count = 0;
        self.mutex.unlock();
        messageFreeList(head, self.allocator);
    }

    pub fn isConnected(self: *const Self) bool {
        return self.connected;
    }

    // ── Mosquitto C callbacks ────────────────────────────────────────

    fn rdrOnConnect(mosq: ?*c.struct_mosquitto, userdata: ?*anyopaque, rc: c_int) callconv(.C) void {
        _ = mosq;
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.mutex.lock();
        defer self.mutex.unlock();
        self.connected = (rc == 0);
        log.info("[reader] on_connect rc={d}", .{rc});
        self.connect_cond.signal();
    }

    fn rdrOnDisconnect(mosq: ?*c.struct_mosquitto, userdata: ?*anyopaque, rc: c_int) callconv(.C) void {
        _ = mosq;
        const self: *Self = @ptrCast(@alignCast(userdata));
        self.mutex.lock();
        defer self.mutex.unlock();
        self.connected = false;
        log.info("[reader] on_disconnect rc={d}", .{rc});
        self.connect_cond.signal();
    }

    fn rdrOnMessage(mosq: ?*c.struct_mosquitto, userdata: ?*anyopaque, msg: ?*const c.struct_mosquitto_message) callconv(.C) void {
        _ = mosq;
        const self: *Self = @ptrCast(@alignCast(userdata));
        const m = msg orelse return;

        // Allocate a new Message node
        const node = self.allocator.create(Message) catch return;
        const topic_ptr: [*]const u8 = @ptrCast(m.topic orelse return);
        const topic_len = std.mem.len(@as([*:0]const u8, @ptrCast(m.topic orelse return)));
        node.topic = self.allocator.alloc(u8, topic_len) catch {
            self.allocator.destroy(node);
            return;
        };
        @memcpy(node.topic, topic_ptr[0..topic_len]);

        const payload_ptr: [*]const u8 = @ptrCast(m.payload orelse {
            self.allocator.free(node.topic);
            self.allocator.destroy(node);
            return;
        });
        const payload_len: usize = @intCast(m.payloadlen);
        node.payload = self.allocator.alloc(u8, payload_len) catch {
            self.allocator.free(node.topic);
            self.allocator.destroy(node);
            return;
        };
        @memcpy(node.payload, payload_ptr[0..payload_len]);
        node.next = null;

        // Append to linked list
        self.mutex.lock();
        defer self.mutex.unlock();
        if (self.msg_tail) |tail| {
            tail.next = node;
        } else {
            self.msg_head = node;
        }
        self.msg_tail = node;
        self.msg_count += 1;
    }

    fn rdrOnSubscribe(
        mosq: ?*c.struct_mosquitto,
        userdata: ?*anyopaque,
        mid: c_int,
        _: c_int,
        _: ?*const c_int,
    ) callconv(.C) void {
        _ = mosq;
        const self: *Self = @ptrCast(@alignCast(userdata));
        log.info("[reader] on_subscribe mid={d}", .{mid});
        self.mutex.lock();
        defer self.mutex.unlock();
        if (mid == self.subscribe_mid) {
            self.subscribe_ack = true;
            self.subscribe_cond.signal();
        }
    }
};

// ──────────────────────────────────────────────────────────────────────
//  Tests
// ──────────────────────────────────────────────────────────────────────

test "Config defaults" {
    const cfg = Config{};
    try std.testing.expectEqualStrings("localhost", cfg.host);
    try std.testing.expectEqual(@as(u16, 1883), cfg.port);
    try std.testing.expect(cfg.clean_session);
    try std.testing.expectEqual(@as(?[:0]const u8, null), cfg.client_id);
}

test "messageFreeList handles null" {
    messageFreeList(null, std.testing.allocator);
}

test "Message allocate and free" {
    const allocator = std.testing.allocator;

    const node = try allocator.create(Message);
    node.topic = try allocator.dupe(u8, "test/topic");
    node.payload = try allocator.dupe(u8, "{\"hello\":1}");
    node.next = null;

    messageFreeList(node, allocator);
}